### ISSUE TITLE
Possible unobserved IOException in HttpClient using SocketsHttpHandler

### ISSUE THREAD

**Issue by Bluezen** [dotnet/runtime#105154]
### Description

While investigating unobserved task exceptions captured by the `UnobservedTaskException` event handler in one of our production applications, we encountered one with a detailed stack trace. 
The exception is `System.IO.IOException: Unable to read data from the transport connection: Connection reset by peer`, and it was triggered by the `SendAsync` method of `HttpClient`. Although we are using a custom `DelegatingHandler`, it doesn't seem relevant to the current issue, as the exception occurs in the `SocketsHttpHandler.SendAsync` method, before our custom `DelegatingHandler` code executes. 

Here is what the stack trace looks like:
```bash
System.AggregateException: A Task's exception(s) were not observed either by Waiting on the Task or accessing its Exception property. As a result, the unobserved exception was rethrown by the finalizer thread. (Unable to read data from the transport connection: Connection reset by peer.)
 ---> System.IO.IOException: Unable to read data from the transport connection: Connection reset by peer.
 ---> System.Net.Sockets.SocketException (104): Connection reset by peer
   at System.Net.Sockets.NetworkStream.ReadAsync(Memory`1 buffer, CancellationToken cancellationToken)
   at System.Net.Http.HttpConnection.PrepareForReuse(Boolean async)
   at System.Net.Http.HttpConnectionPool.TryGetPooledHttp11Connection(HttpRequestMessage request, Boolean async, HttpConnection& connection, HttpConnectionWaiter`1& waiter)
   at System.Net.Http.HttpConnectionPool.SendWithVersionDetectionAndRetryAsync(HttpRequestMessage request, Boolean async, Boolean doRequestAuth, CancellationToken cancellationToken)
   at System.Runtime.CompilerServices.AsyncMethodBuilderCore.Start[TStateMachine](TStateMachine& stateMachine)
   at System.Net.Http.HttpConnectionPool.SendAsync(HttpRequestMessage request, Boolean async, Boolean doRequestAuth, CancellationToken cancellationToken)
   at System.Net.Http.HttpConnectionPoolManager.SendAsyncCore(HttpRequestMessage request, Uri proxyUri, Boolean async, Boolean doRequestAuth, Boolean isProxyConnect, CancellationToken cancellationToken)
   at System.Net.Http.HttpConnectionPoolManager.SendAsync(HttpRequestMessage request, Boolean async, Boolean doRequestAuth, CancellationToken cancellationToken)
   at System.Net.Http.DiagnosticsHandler.SendAsyncCore(HttpRequestMessage request, Boolean async, CancellationToken cancellationToken)
   at System.Runtime.CompilerServices.AsyncMethodBuilderCore.Start[TStateMachine](TStateMachine& stateMachine)
   at System.Net.Http.HttpMessageHandlerStage.SendAsync(HttpRequestMessage request, CancellationToken cancellationToken)
   at System.Net.Http.SocketsHttpHandler.SendAsync(HttpRequestMessage request, CancellationToken cancellationToken)
   at MyAwesomeCompany.OurCustomDelegatingHandler.SendAsync(HttpRequestMessage request, CancellationToken cancellationToken) in /builds/OurCustomDelegatingHandler.cs:line 18
   at System.Net.Http.HttpClient.<SendAsync>g__Core|83_0(HttpRequestMessage request, HttpCompletionOption completionOption, CancellationTokenSource cts, Boolean disposeCts, CancellationTokenSource pendingRequestsCts, CancellationToken originalCancellationToken)

```
The stack trace reveals that a `ReadAsync` call on a stream in the `HttpConnection.PrepareForReuse` method triggers the `IOException`. Looking at the `HttpConnection` code, it appears to occur at the [following line](https://source.dot.net/#System.Net.Http/System/Net/Http/SocketsHttpHandler/HttpConnection.cs,170):
```csharp
_readAheadTask = _stream.ReadAsync(_readBuffer.AvailableMemory);
```

I am aware that `HttpConnection`'s code has been updated to address a similar issue with the [following fix](https://github.com/dotnet/runtime/pull/104335). 
The goal of this ticket is to discuss whether my issue should be resolved with this fix and to consider the possibility that the problem could still occur even after the fix.

### Reproduction Steps

I don't have a way to reproduce the exact issue, but here is a small code snippet (which you [can try here](https://dotnetfiddle.net/K5OyRI)) that attempts to simulate/replicate the behavior of the issue within the [HttpConnection](https://source.dot.net/System.Net.Http/R/b7e9b61a384f3eb0.html)'s code :
```csharp
using System;
using System.Threading.Tasks;
using System.Net;
using System.Net.Http;
using System.Threading;
using System.Net.Sockets;

public class Program
{
    public static async Task Main()
    {
        // Arrange
        TaskScheduler.UnobservedTaskException += TaskSchedulerOnUnobservedTaskException;

        using var handler = new FaillingSocketHttpHandler();
        using var client = new HttpClient(handler);

        using var request = new HttpRequestMessage(HttpMethod.Get, "http://example.com");

        // Act
        HttpResponseMessage response = await client.SendAsync(request);

        // Wait a bit for the task to complete, as the exception is thrown in a different thread which is not awaited
        await Task.Delay(100);

        // Force garbage collection and wait for finalizers to run to ensure the exception is caught by TaskScheduler.UnobservedTaskException
        handler.Task =
            default; // Ensure the task will be collected. Similarly to what happens here: https://source.dot.net/#System.Net.Http/System/Net/Http/SocketsHttpHandler/HttpConnection.cs,620
        GC.Collect();
        GC.WaitForPendingFinalizers();

        // Print response
        Console.WriteLine($"Response received: {await response.Content.ReadAsStringAsync()}");

        // Cleanup
        TaskScheduler.UnobservedTaskException -= TaskSchedulerOnUnobservedTaskException;
        return;

        void TaskSchedulerOnUnobservedTaskException(object sender, UnobservedTaskExceptionEventArgs e)
        {
            if (e.Exception.InnerException is SocketException socketException &&
                socketException.SocketErrorCode == SocketError.ConnectionReset)
            {
                Console.WriteLine(
                    $"The SocketException was caught by TaskScheduler.UnobservedTaskException: {e.Exception.InnerException.Message}");
            }
        }
    }
}

// This class simulates a scenario where a SocketException occurs unexpectedly during the execution of SendAsync,
// similar to the behavior observed in System.Net.Http.HttpConnection's PrepareForReuse method.
// Reference: https://source.dot.net/#System.Net.Http/System/Net/Http/SocketsHttpHandler/HttpConnection.cs,169
public class FaillingSocketHttpHandler : HttpMessageHandler
{
    public Task Task { get; set; }

    protected override async Task<HttpResponseMessage> SendAsync(HttpRequestMessage request,
        CancellationToken cancellationToken)
    {
        try
        {
            // This task mimics the exception thrown by NetworkStream.ReadAsync in the PrepareForReuse method,
            // which can occur if a socket is reset immediately after a task completion check.
#pragma warning disable CA2012 // we reproduce the same behavior as the PrepareForReuse method
            Task = Task.Run(() =>
            {
                // /!\ Mimic NetworkStream.ReadAsync asynchronous operation by waiting just enough so that
                // Task.IsCompleted check below does not return true
                Thread.Sleep(20);
                // Then we simulate the exception thrown by the socket
                throw new SocketException((int)SocketError.ConnectionReset);
            });
#pragma warning restore CA2012
            if (Task.IsCompleted) // Replicate the behavior of the PrepareForReuse method
            {
                Console.WriteLine("The task should not be completed yet");
                return new HttpResponseMessage(HttpStatusCode.Gone);
            }
        }
        catch (Exception e) // Replicate the behavior of the PrepareForReuse method
        {
            Console.WriteLine($"Unexpected exception: {e}");
            return new HttpResponseMessage(HttpStatusCode.BadRequest);
        }

        await Task.Delay(1, cancellationToken); // Not required, just to ensure the task is not completed synchronously

        var response = new HttpResponseMessage(HttpStatusCode.OK)
        {
            RequestMessage = request,
            Content = new StringContent("Hi from FaillingSocketHttpHandler!")
        };
        return response;
    }
}
```

### Expected behavior

The read-ahead task should avoid leaking unobserved exceptions. 

### Actual behavior

The read-ahead operation throws an exception that is not caught or observed. The finalizer thread then rethrows the unobserved exception, which is subsequently captured by the UnobservedTaskException event handler.

### Configuration

The application is compiled for x64 and running on Docker bare metal with official AspNet image with .NET in version 8.0.7.

### Other information

I believe `PrepareForReuse` still has an issue with handling unobserved exceptions that the `ReadAsync` operation could throw if the task doesn't complete synchronously. More consideration is needed, but if it's not possible to await the `ReadAsync` method, a less ideal alternative could be to add a continuation to `_readAheadTask` with `.ContinueWith(t => LogExceptions(t), TaskContinuationOptions.OnlyOnFaulted)` in an `else` condition of the existing `if(_readAheadTask.IsCompleted)` check.

**Comment by MihaZupan** [comment]
https://github.com/dotnet/runtime/pull/104335 should indeed fix this issue for 9.0.

> PrepareForReuse still has an issue with handling unobserved exceptions that the ReadAsync operation could throw if the task doesn't complete synchronously.

If the `ReadAsync` doesn't complete right away, `PrepareForReuse` will return true. The contract of the method is such that if it returns true, the caller must follow up with a call to `SendAsync`, where we're guaranteed to observe the readAhead task.

**Comment by rchoffardet** [comment]
Could a backport to 8.0 be considered or would the fix only be available when 9.0 goes GA?

**Comment by ManickaP** [comment]
The fix will be only in 9.0. This doesn't meet the bar for backport.

**Comment by ManickaP** [comment]
Duplicate of #104324

**Comment by Bluezen** [comment]
@MihaZupan thanks for your reply! 
@ManickaP , apologies for the delay in responding, but I would like to follow up on @MihaZupan's answer before closing this issue. As I mentioned in my initial message, I am aware of the fix provided in [PR #104335](https://github.com/dotnet/runtime/pull/104335) for [issue #104324](https://github.com/dotnet/runtime/issues/104324), but I think the code might still have issues handling unobserved exceptions.
> > PrepareForReuse still has an issue with handling unobserved exceptions that the ReadAsync operation could throw if the task doesn't complete synchronously.
> 
> If the `ReadAsync` doesn't complete right away, `PrepareForReuse` will return true. The contract of the method is such that if it returns true, the caller must follow up with a call to `SendAsync`, where we're guaranteed to observe the readAhead task.

Please correct me if I am mistaken, but looking at the `HttpConnection` [code](https://github.com/dotnet/runtime/blob/main/src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnection.cs#L532) from `SocketsHttpHandler`, it seems there are multiple paths that could lead to exceptions not to be observe.  My code snippet was an attempt to reproduce the issue, but it might be clearer if I annotate the extracts of the code directly with comments to point out the potential issues:
```csharp
        public bool PrepareForReuse(bool async)
        {
            (...)
            if (!async && _stream is NetworkStream networkStream)
            {
                (...)
            }
            else
            {
                (...)
                try
                {
#pragma warning disable CA2012
                    _readAheadTask = _stream.ReadAsync(_readBuffer.AvailableMemory);
#pragma warning restore CA2012

                   (...)
                    if (_readAheadTask.IsCompleted)
                    {
                        LogExceptions(_readAheadTask.AsTask());
                        return false;
                    }
                    // the read task could complete and throw an exception here
                    return true;
                }
                catch (Exception error)
                {
                    (...)
                    // This catch will NEVER eat a reading error because the read task is not awaited
                }
            }
        }

        public async Task<HttpResponseMessage> SendAsync(HttpRequestMessage request, bool async, CancellationToken cancellationToken)
        {
            (...)
            try
            {
                (...)
                if (ReadAheadTaskHasStarted)
                {
                    (...)
                    ValueTask<int> vt = _readAheadTask;
                    _readAheadTask = default;

                    int bytesRead;
                    if (vt.IsCompleted)
                    {
                        bytesRead = vt.Result;
                       // vt could be faulted here and have an unobserved exception
                    }
                    else
                    {
                        (...)
                        bytesRead = await vt.ConfigureAwait(false);
                    }
                    (...)
                }
                (...)
            }
            catch (Exception error)
            {
                (...)
                if (_readAheadTask != default)
                {
                    (...)
                    LogExceptions(_readAheadTask.AsTask());
                }
                (...)
            }
        }
```

**Comment by MihaZupan** [comment]
> // the read task could complete and throw an exception here
> return true;

This is fine per my previous comment - if we return true, the caller will observe the task in `SendAsync`.

> // This catch will NEVER eat a reading error because the read task is not awaited

If the call to `ReadAsync` synchronously throws, there is no `Task`, and therefore no unobserved exceptions.

> // vt could be faulted here and have an unobserved exception

Calling `.Result` right before will observe and rethrow any such exception.

Let me know if I've missed something.

**Comment by Bluezen** [comment]
Alright, I think you are right on all points. Thank you for your time. We will wait for the release of version 9.0.
